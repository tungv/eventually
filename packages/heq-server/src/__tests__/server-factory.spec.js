const enableDestroy = require('server-destroy');
const got = require('got');
const http = require('http');
const ports = require('port-authority');

const factory = require('../factory');

const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));

const commitSomething = async ({ port }) => {
  const { body } = await got(`http://localhost:${port}/commit`, {
    json: true,
    body: {
      type: 'TEST',
      payload: {
        key: 'value',
      },
    },
  });

  return body;
};

const simpleParse = buffer => {
  const chunks = buffer.split('\n\n').filter(x => x);

  if (chunks[0] !== ':ok') {
    throw new Error('not ok');
  }

  const events = chunks.map(ch => {
    const lines = ch.split('\n');
    if (lines[1] !== 'event: INCMSG') {
      return [];
    }

    return JSON.parse(lines[2].slice('data: '.length));
  });

  return [].concat(...events);
};

describe('factory()', () => {
  it('should return start function', async () => {
    const port = await ports.find(30000);
    const { start } = await factory({
      http: { port },
    });

    const server = await start();
    server.close();

    expect(server).toBeInstanceOf(http.Server);
  });

  it('should able to commit', async () => {
    const port = await ports.find(30000);
    const { start } = await factory({
      http: { port },
    });

    const server = await start();
    enableDestroy(server);

    const { body } = await got(`http://localhost:${port}/commit`, {
      json: true,
      body: {
        type: 'TEST',
        payload: {
          key: 'value',
        },
      },
    });

    server.destroy();

    expect(body).toEqual({ id: 1, payload: { key: 'value' }, type: 'TEST' });
  });

  it('should able to query', async () => {
    const port = await ports.find(30000);
    const { start } = await factory({
      http: { port },
    });

    const server = await start();
    enableDestroy(server);

    for (let i = 0; i < 5; ++i) await commitSomething({ port });

    const { body } = await got(`http://localhost:${port}/query`, {
      json: true,
      query: {
        lastEventId: 2,
      },
    });

    server.destroy();

    expect(body).toEqual([
      { id: 3, payload: { key: 'value' }, type: 'TEST' },
      { id: 4, payload: { key: 'value' }, type: 'TEST' },
      { id: 5, payload: { key: 'value' }, type: 'TEST' },
    ]);
  });

  it('should able to subscribe', async done => {
    const port = await ports.find(30000);
    const { start } = await factory({
      http: { port },
    });

    const server = await start();
    enableDestroy(server);

    for (let i = 0; i < 5; ++i) await commitSomething({ port });

    const stream = got.stream(`http://localhost:${port}/subscribe`, {
      headers: {
        'Last-Event-ID': 2,
        'Burst-Time': 1,
      },
    });

    const buffer = [];

    stream.on('data', data => {
      const msg = String(data);

      buffer.push(msg);

      // stop after receiving event #5
      if (msg.slice(0, 5) === 'id: 5') {
        server.destroy();
      }
    });

    stream.on('end', () => {
      expect(simpleParse(buffer.join(''))).toEqual([
        { id: 3, payload: { key: 'value' }, type: 'TEST' },
        { id: 4, payload: { key: 'value' }, type: 'TEST' },
        { id: 5, payload: { key: 'value' }, type: 'TEST' },
      ]);
      done();
    });
  });

  it('should able to subscribe from a past event id', async done => {
    const port = await ports.find(30000);
    const { start } = await factory({
      http: { port },
    });

    const server = await start();
    enableDestroy(server);

    for (let i = 0; i < 5; ++i) await commitSomething({ port });

    const stream = got.stream(`http://localhost:${port}/subscribe`, {
      headers: {
        'Last-Event-ID': 2,
        'Burst-Time': 1,
      },
    });

    for (let i = 0; i < 5; ++i) await commitSomething({ port });

    const buffer = [];

    stream.on('data', data => {
      const msg = String(data);

      buffer.push(msg);

      // stop after receiving event #5

      const chunks = msg.split('\n\n');

      for (const chunk of chunks) {
        if (chunk.slice(0, 6) === 'id: 10') {
          server.destroy();
        }
      }
    });

    stream.on('end', () => {
      const events = simpleParse(buffer.join(''));

      expect(events).toEqual([
        { id: 3, payload: { key: 'value' }, type: 'TEST' },
        { id: 4, payload: { key: 'value' }, type: 'TEST' },
        { id: 5, payload: { key: 'value' }, type: 'TEST' },
        { id: 6, payload: { key: 'value' }, type: 'TEST' },
        { id: 7, payload: { key: 'value' }, type: 'TEST' },
        { id: 8, payload: { key: 'value' }, type: 'TEST' },
        { id: 9, payload: { key: 'value' }, type: 'TEST' },
        { id: 10, payload: { key: 'value' }, type: 'TEST' },
      ]);
      done();
    });
  });

  it('should fallback to documentation page instead of 404', async () => {
    const port = await ports.find(40000);
    const { start } = await factory({
      http: { port },
    });

    const server = await start();
    enableDestroy(server);

    const { body } = await got(`http://localhost:${port}/`);

    server.destroy();

    expect(body).toMatchSnapshot('documentation content');
  });
});
