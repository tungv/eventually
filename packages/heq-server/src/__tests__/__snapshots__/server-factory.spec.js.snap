// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factory() should fallback to documentation page instead of 404: documentation content 1`] = `"<title>heq-server hint page</title><body><h3>commit</h3><ol><li><code><pre>curl -XPOST http://127.0.0.1:40000/commit \\\\</pre></code></li><li><code><pre>     -d'{\\"type\\": \\"<event-name>\\", \\"payload\\": {\\"key\\": \\"value\\"}}' \\\\</pre></code></li><li><code><pre>     -H\\"content-type: application/json\\"</pre></code></li></ol><h3>query from event #101 - now</h3><ol><li><code><pre>curl http://127.0.0.1:40000/query?lastEventId=100</pre></code></li></ol><h3>subscribe from event #101 - now</h3><ol><li><code><pre>http://127.0.0.1:40000/subscribe \\\\</pre></code></li><li><code><pre>-H\\"Last-Event-ID: 100\\"</pre></code></li></ol></body>"`;
